src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -480,6 +461,108 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
             _incomingBuffer.Discard(frameHeader.Length);
         }
 
+        private async void SendSettingsAck()
+        {
+            await _writerLock.WaitAsync().ConfigureAwait(false);
+            try
+            {
+                WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
+
+                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+            }
+            finally
+            {
+                _writerLock.Release();
+            }
+        }
+
+        private async void SendPingAck(ReadOnlyMemory<byte> pingContent)
+        {
+            Debug.Assert(pingContent.Length == FrameHeader.PingLength);
+
+            await _writerLock.WaitAsync().ConfigureAwait(false);
+            try
+            {
+                WriteFrameHeader(new FrameHeader(FrameHeader.PingLength, FrameType.Ping, FrameFlags.Ack, 0));
+                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.PingLength);
+                pingContent.CopyTo(_outgoingBuffer.AvailableMemory);
+                _outgoingBuffer.Commit(FrameHeader.PingLength);
+
+                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+            }
+            finally
+            {
+                _writerLock.Release();
+            }
+        }
+
+        private async void SendRstStream(int streamId, Http2ProtocolErrorCode errorCode)
+        {
+            await _writerLock.WaitAsync().ConfigureAwait(false);
+            try
+            {
+                WriteFrameHeader(new FrameHeader(FrameHeader.RstStreamLength, FrameType.RstStream, FrameFlags.None, streamId));
+
+                _outgoingBuffer.AvailableSpan[0] = (byte)(((int)errorCode & 0xFF000000) >> 24);
+                _outgoingBuffer.AvailableSpan[1] = (byte)(((int)errorCode & 0x00FF0000) >> 16);
+                _outgoingBuffer.AvailableSpan[2] = (byte)(((int)errorCode & 0x0000FF00) >> 8);
+                _outgoingBuffer.AvailableSpan[3] = (byte)((int)errorCode & 0x000000FF);
+                _outgoingBuffer.Commit(FrameHeader.RstStreamLength);
+
+                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+            }
+            finally
+            {
+                _writerLock.Release();
+            }
+        }
+
+        private (ReadOnlyMemory<byte> first, ReadOnlyMemory<byte> rest) SplitBufferForFraming(ReadOnlyMemory<byte> buffer) =>
+            buffer.Length > FrameHeader.MaxLength ?
+                (buffer.Slice(0, FrameHeader.MaxLength), buffer.Slice(FrameHeader.MaxLength)) :
+                (buffer, Memory<byte>.Empty);
+
+        private async ValueTask SendStreamDataAsync(int streamId, ReadOnlyMemory<byte> buffer)
```

Comment:
Do you have a plan for how this is going to work with window size limitations from the server?

Reply:
Yes. Basically, we need to track the current window size and, if it's gone to 0, we'll await more window being available. We will do this per frame. I have some old code that implements this, but I don't have it implemented in the new refactoring here. I'll add it in a subsequent PR.